locations/jclouds/src/main/java/org/apache/brooklyn/location/jclouds/JcloudsLocation.java

```diff
@@ -2862,15 +2863,61 @@ protected String getFirstReachableAddress(NodeMetadata node, ConfigBag setup) {
+                        + " : " + setup.get(POLL_FOR_FIRST_REACHABLE_ADDRESS_PREDICATE_TYPE));
+
+                List<Constructor<?>> constructors = Arrays.asList(setup.get(POLL_FOR_FIRST_REACHABLE_ADDRESS_PREDICATE_TYPE).getConstructors());
+                Optional<Constructor<?>> emptyConstructor= Iterables.tryFind(constructors, new Predicate<Constructor<?>>() {
+                    @Override public boolean apply(Constructor<?> constructor) {
+                        return constructor.getParameterTypes().length == 0;
+                    }
+                });
+                Optional<Constructor<?>> mapConstructor = Iterables.tryFind(constructors, new Predicate<Constructor<?>>() {
+                    @Override public boolean apply(Constructor<?> constructor) {
+                        if (constructor.getParameterTypes().length == 1) {
+                            return constructor.getParameterTypes()[0].isAssignableFrom(Map.class);
+                        } else {
+                            return false;
+                        }
+                    }
+                });
+                if (mapConstructor.isPresent()) {
+                    Map<String, Object> args = MutableMap.of();
+                    ConfigUtils.addUnprefixedConfigKeyInConfigBack(POLL_FOR_FIRST_REACHABLE_ADDRESS_PREDICATE.getName() + ".", setup, args);
+                    try {
+                        pollForFirstReachableHostAndPortPredicate = setup.get(POLL_FOR_FIRST_REACHABLE_ADDRESS_PREDICATE_TYPE).getConstructor(Map.class).newInstance(args);
+                    } catch (ReflectiveOperationException e) {
+                        throw Exceptions.propagate(e);
+                    }
+                } else {
+                    try {
+                        pollForFirstReachableHostAndPortPredicate = (Predicate<HostAndPort>)emptyConstructor.get().newInstance();
+                    } catch (ReflectiveOperationException e) {
+                        throw Exceptions.propagate(e);
+                    }
+                }
+            }
+
+            try {
+                result = JcloudsUtil.getFirstReachableAddress(node, timeout, pollForFirstReachableHostAndPortPredicate);
+            } catch (Exception e) {
+                throw Exceptions.propagate("Problem instantiating reachability checker for " + this
+                        + " pollForFirstReachableHostAndPortPredicate: " + pollForFirstReachableHostAndPortPredicate, e);
+            }
         } else {
             result = Iterables.getFirst(Iterables.concat(node.getPublicAddresses(), node.getPrivateAddresses()), null);
             if (result == null) {
                 throw new IllegalStateException("No addresses available for node "+node+" in "+this);
             }
-            LOG.debug("Using first address "+result+" for node "+node+" in "+this);
```

Comment:
This is an important distinction for troubleshooting. Leave both the first and first-reachable log statements.